File: api/routes/chartink.py
from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from api.services.chartink_service import (
    create_strategy,
    get_strategy,
    list_strategies,
    update_strategy,
    delete_strategy,
    process_webhook,
    list_symbol_mappings,
    add_symbol_mapping,
    remove_symbol_mapping,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/openalgo/chartink", tags=["chartink"])


class StrategyCreate(BaseModel):
    name: str
    symbol: str
    exchange: str = "NSE"
    action: str = "BUY"
    quantity: int = 1
    product: str = "MIS"
    pricetype: str = "MARKET"
    intraday: bool = True
    enabled: bool = True


class StrategyUpdate(BaseModel):
    name: str | None = None
    symbol: str | None = None
    exchange: str | None = None
    action: str | None = None
    quantity: int | None = None
    product: str | None = None
    pricetype: str | None = None
    intraday: bool | None = None
    enabled: bool | None = None


class SymbolMappingCreate(BaseModel):
    chartink_symbol: str
    trading_symbol: str
    exchange: str = "NSE"


@router.post("/webhook")
async def webhook(body: dict):
    result = await process_webhook(body)
    return result


@router.get("/strategies")
async def get_strategies():
    return list_strategies()


@router.post("/strategies")
async def create_strategy_endpoint(data: StrategyCreate):
    return create_strategy(data.model_dump())


@router.get("/strategies/{id}")
async def get_strategy_endpoint(id: str):
    strategy = get_strategy(id)
    if not strategy:
        raise HTTPException(status_code=404, detail="Strategy not found")
    return strategy


@router.put("/strategies/{id}")
async def update_strategy_endpoint(id: str, data: StrategyUpdate):
    strategy = update_strategy(id, data.model_dump(exclude_none=True))
    if not strategy:
        raise HTTPException(status_code=404, detail="Strategy not found")
    return strategy


@router.delete("/strategies/{id}")
async def delete_strategy_endpoint(id: str):
    if not delete_strategy(id):
        raise HTTPException(status_code=404, detail="Strategy not found")
    return {"status": "deleted"}


@router.get("/symbols")
async def get_symbols():
    return list_symbol_mappings()


@router.post("/symbols")
async def create_symbol_mapping(data: SymbolMappingCreate):
    return add_symbol_mapping(data.model_dump())


@router.delete("/symbols/{id}")
async def delete_symbol_mapping(id: str):
    if not remove_symbol_mapping(id):
        raise HTTPException(status_code=404, detail="Symbol mapping not found")
    return {"status": "deleted"}

File: frontend/src/api/chartink.ts
import { api } from './client'
import type { ChartinkStrategy, ChartinkSymbolMapping } from '../types/chartink'

export async function fetchChartinkStrategies(): Promise<ChartinkStrategy[]> {
  const { data } = await api.get('/openalgo/chartink/strategies')
  return data
}

export async function fetchChartinkStrategy(id: string): Promise<ChartinkStrategy> {
  const { data } = await api.get(`/openalgo/chartink/strategies/${id}`)
  return data
}

export async function createChartinkStrategy(payload: {
  name: string
  symbol: string
  exchange: string
  action: string
  quantity: number
  product: string
  pricetype: string
  intraday: boolean
}): Promise<ChartinkStrategy> {
  const { data } = await api.post('/openalgo/chartink/strategies', payload)
  return data
}

export async function updateChartinkStrategy(id: string, payload: Partial<ChartinkStrategy>): Promise<ChartinkStrategy> {
  const { data } = await api.put(`/openalgo/chartink/strategies/${id}`, payload)
  return data
}

export async function deleteChartinkStrategy(id: string): Promise<void> {
  await api.delete(`/openalgo/chartink/strategies/${id}`)
}

export async function fetchChartinkSymbolMappings(): Promise<ChartinkSymbolMapping[]> {
  const { data } = await api.get('/openalgo/chartink/symbols')
  return data
}

export async function createChartinkSymbolMapping(payload: {
  chartink_symbol: string
  trading_symbol: string
  exchange: string
}): Promise<ChartinkSymbolMapping> {
  const { data } = await api.post('/openalgo/chartink/symbols', payload)
  return data
}

export async function deleteChartinkSymbolMapping(id: string): Promise<void> {
  await api.delete(`/openalgo/chartink/symbols/${id}`)
}

export async function sendChartinkWebhook(payload: Record<string, unknown>): Promise<unknown> {
  const { data } = await api.post('/openalgo/chartink/webhook', payload)
  return data
}

export async function sendTradingViewWebhook(payload: Record<string, unknown>): Promise<unknown> {
  const { data } = await api.post('/openalgo/tradingview/webhook', payload)
  return data
}

export async function sendGoChartingWebhook(payload: Record<string, unknown>): Promise<unknown> {
  const { data } = await api.post('/openalgo/gocharting/webhook', payload)
  return data
}
